### PR TITLE
Grid halo swap

### PR DESCRIPTION
**Pull request by matt-frey**

### CODE REVIEW

File: src/3d/fields/field_mpi.f90 [hunk 1]
```diff
@@ -1,4 +1,128 @@
 module field_mpi
+    use mpi_layout
+    use mpi_communicator
     implicit none
 
+    integer, parameter :: HALO_WEST_TAG  = 100000,   &
+                          HALO_EAST_TAG  = 100001,   &
+                          HALO_SOUTH_TAG = 100002,   &
+                          HALO_NORTH_TAG = 100003,   &
+                          HALO_LOWLE_TAG = 100004,   &
```
Comment (sjboeing) [inline src/3d/fields/field_mpi.f90:10]: Should we use North, West, northwest etc throughout for consistency?

File: src/mpi/mpi_layout.f90 [hunk 2]
```diff
@@ -23,20 +23,22 @@ module mpi_layout
...
-                neighbour%right = mpi_rank
+                neighbour%west  = mpi_rank
+                neighbour%east = mpi_rank
                 neighbour%south = mpi_rank
                 neighbour%north = mpi_rank
-                box%lo = (/0,  0,  0 /)
-                box%hi = (/nx, ny, nz/)
+                box%lo = (/0,  0,  -1  /)
+                box%hi = (/nx, ny, nz+1/)
+                box%hlo = box%lo
```
Comment (sjboeing) [inline src/mpi/mpi_layout.f90:40]: Is this needed?

Comment (matt-frey) [reply src/mpi/mpi_layout.f90:40]: Yes, it is.

File: src/3d/fields/field_mpi.f90 (hunk 1 above)
Comment (matt-frey) [reply src/3d/fields/field_mpi.f90:10]: I am not yet happy with the naming North, West etc.

Comment (matt-frey) [reply src/3d/fields/field_mpi.f90:10]: Not sure if left, right, top, bottom is better.

Comment (matt-frey) [reply src/3d/fields/field_mpi.f90:10]: In PMPIC you use left, right, up and down.

Comment (matt-frey) [reply src/3d/fields/field_mpi.f90]: Let's keep north, west etc.

File: src/3d/fields/field_mpi.f90 [hunk 3]
```diff
@@ -1,4 +1,447 @@
...
+
+        end subroutine field_halo_reset
+
+
+        subroutine field_halo_fill(data)
+            double precision, intent(inout) :: data(box%hlo(3):box%hhi(3), &
+                                                    box%hlo(2):box%hhi(2), &
+                                                    box%hlo(1):box%hhi(1))
+
+            if (mpi_size == 1) then
```
Comment (sjboeing) [inline src/3d/fields/field_mpi.f90:60]: @matt-frey: I was wondering about this line (realise it has been merged by now). Doesn't "mpi_size ==1" also need filled halos for derivatives?

Comment (sjboeing) [reply src/3d/fields/field_mpi.f90:60]: Just looking at the PR, think you did a really nice job from what I see so far.

Comment (matt-frey) [reply src/3d/fields/field_mpi.f90:60]: @sjboeing As discussed I remove this if-statement with b62bcd8125f7b005caf80bc9330a90532977269e.

**Pull request merged**